crates/wasi-http/src/struct.rs

```diff
@@ -1,124 +1,391 @@
+    }
+    fn delete_request(&mut self, id: u32) -> Result<(), TableError> {
+        self.delete::<Box<dyn HttpRequest>>(id).map(|_old| ())
+    }
+
+    fn push_response(&mut self, response: Box<dyn HttpResponse>) -> Result<u32, TableError> {
+        self.push(Box::new(response))
+    }
+    fn get_response(&self, id: u32) -> Result<&dyn HttpResponse, TableError> {
+        self.get::<Box<dyn HttpResponse>>(id).map(|f| f.as_ref())
+    }
+    fn get_response_mut(&mut self, id: u32) -> Result<&mut Box<dyn HttpResponse>, TableError> {
+        self.get_mut::<Box<dyn HttpResponse>>(id)
+    }
+    fn delete_response(&mut self, id: u32) -> Result<(), TableError> {
+        self.delete::<Box<dyn HttpResponse>>(id).map(|_old| ())
+    }
+
+    fn push_future(&mut self, future: Box<ActiveFuture>) -> Result<u32, TableError> {
+        self.push(Box::new(future))
+    }
+    fn get_future(&self, id: u32) -> Result<&ActiveFuture, TableError> {
+        self.get::<Box<ActiveFuture>>(id).map(|f| f.as_ref())
+    }
+    fn get_future_mut(&mut self, id: u32) -> Result<&mut Box<ActiveFuture>, TableError> {
+        self.get_mut::<Box<ActiveFuture>>(id)
+    }
+    fn delete_future(&mut self, id: u32) -> Result<(), TableError> {
+        self.delete::<Box<ActiveFuture>>(id).map(|_old| ())
+    }
+
+    fn push_fields(&mut self, fields: Box<ActiveFields>) -> Result<u32, TableError> {
+        self.push(Box::new(fields))
+    }
+    fn get_fields(&self, id: u32) -> Result<&ActiveFields, TableError> {
+        self.get::<Box<ActiveFields>>(id).map(|f| f.as_ref())
+    }
+    fn get_fields_mut(&mut self, id: u32) -> Result<&mut Box<ActiveFields>, TableError> {
+        self.get_mut::<Box<ActiveFields>>(id)
+    }
+    fn delete_fields(&mut self, id: u32) -> Result<(), TableError> {
+        self.delete::<Box<ActiveFields>>(id).map(|_old| ())
+    }
+
+    fn push_stream(&mut self, content: Bytes, parent: u32) -> Result<(u32, Stream), TableError> {
+        let (a, b) = tokio::io::duplex(MAX_BUF_SIZE);
+        let (_, write_stream) = tokio::io::split(a);
+        let (read_stream, _) = tokio::io::split(b);
+        let input_stream = AsyncReadStream::new(read_stream);
+        let mut output_stream = AsyncWriteStream::new(write_stream);
+
+        let mut cursor = 0;
+        while cursor < content.len() {
+            let (written, _) = output_stream
+                .write(content.slice(cursor..content.len()))
+                .map_err(|_| TableError::NotPresent)?;
+            cursor += written;
         }
+
+        let input_stream = Box::new(input_stream);
+        let output_id = self.push_output_stream(Box::new(output_stream))?;
+        let input_id = self.push_input_stream(input_stream)?;
+        let stream = Stream::new(input_id, output_id, parent);
+        let cloned_stream = stream.clone();
+        let stream_id = self.push(Box::new(Box::new(stream)))?;
+        Ok((stream_id, cloned_stream))
+    }
```

Comment:
@pchickey , here is the part where we have a shared stream for both read and write. i solved by populating the table with both streams and just keeping their ids for later use.